rpc/grpc/server/services/blockservice/service.go

```diff
@@ -0,0 +1,129 @@
+
+import (
+	context "context"
+	"fmt"
+
+	"github.com/cometbft/cometbft/libs/log"
+	cmtpubsub "github.com/cometbft/cometbft/libs/pubsub"
+	blocksvc "github.com/cometbft/cometbft/proto/tendermint/services/block/v1"
+	"github.com/cometbft/cometbft/store"
+	"github.com/cometbft/cometbft/types"
+	"github.com/google/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type blockServiceServer struct {
+	store    *store.BlockStore
+	eventBus *types.EventBus
+	logger   log.Logger
+}
+
+// New creates a new CometBFT version service server.
+func New(store *store.BlockStore, eventBus *types.EventBus, logger log.Logger) blocksvc.BlockServiceServer {
+	log := logger.With("module", "grpc-block-service")
+	return &blockServiceServer{
+		store,
+		eventBus,
+		log,
+	}
+}
+
+// GetByHeight implements v1.BlockServiceServer GetByHeight method
+func (s *blockServiceServer) GetByHeight(_ context.Context, req *blocksvc.GetByHeightRequest) (*blocksvc.GetByHeightResponse, error) {
+	var height int64
+
+	// retrieve the last height in the store
+	latestHeight := s.store.Height()
+
+	// validate height parameter, if height is 0 or
+	// the request is nil, then use the latest height
+	if req.Height == 0 {
+		height = latestHeight
+	} else if req.Height < 0 {
+		errMsg := fmt.Sprintf("got negative height (%d), please specify a height >= 0", req.Height)
+		s.logger.Error("GetByHeight", "err", errMsg)
+		return nil, status.Error(codes.InvalidArgument, errMsg)
+	} else {
+		height = req.Height
+	}
+
+	// check if the height requested is not higher
+	// than the latest height in the store
+	if height > latestHeight {
+		errMsg := fmt.Sprintf("height requested (%d) is higher than the latest available height (%d)", height, latestHeight)
+		s.logger.Error("GetByHeight", "err", errMsg)
+		return nil, status.Errorf(codes.InvalidArgument, errMsg)
+	}
+
+	block := s.store.LoadBlock(height)
+	blockProto, err := block.ToProto()
+	if err != nil {
+		errMsg := fmt.Sprintf("block at height %d not found", height)
+		s.logger.Error("GetByHeight", "err", errMsg)
+		return nil, status.Errorf(codes.NotFound, errMsg)
+	}
+
+	blockMeta := s.store.LoadBlockMeta(height)
+
+	blockIDProto := blockMeta.BlockID.ToProto()
+
+	return &blocksvc.GetByHeightResponse{
+		BlockId: &blockIDProto,
+		Block:   blockProto,
+	}, nil
+}
+
+// GetLatestHeight implements v1.BlockServiceServer GetLatestHeight method
+func (s *blockServiceServer) GetLatestHeight(_ *blocksvc.GetLatestHeightRequest, stream blocksvc.BlockService_GetLatestHeightServer) error {
+	// Generate a unique subscriber ID using a UUID
+	// The subscriber needs to be unique across all clients
+	id, err := uuid.NewUUID()
+	if err != nil {
+		// cannot generate unique id
+		errMsg := "error generating a subscriber id, cannot subscribe to new block events"
+		s.logger.Error("GetLatestHeight", "err", errMsg)
+		return status.Error(codes.Internal, errMsg)
+	}
+	subscriber := fmt.Sprintf("subscriber_%s", id.String())
```

Comment:
Prefix is redundant.

```suggestion
	subscriber := id.String()
```